### ISSUE TITLE
CIF files

### ISSUE THREAD

**Issue by Luthaf** [Luthaf/Chemharp#10]
File Type                        | text
-------------------------------|---------
Topological information | Yes
Positions                       | Yes
Velocities                      | No

Reference of the foirmat: http://scripts.iucr.org/cgi-bin/paper?S010876739101067X

A library for guessing symmetry is available here: https://github.com/atztogo/spglib

**Comment by Luthaf** [comment]
It appears that CIF files are a written using the STAR (Self-defining Text Archive and Retrieval) file format, using a subset of the STAR rules. The mmCIF is a subset of the CIF files. So implementing CIF reader will give mmCIF for free, and I think that writting CIF files using mmCIF format can not harm.

Relevant links:

* [STAR format paper](http://dx.doi.org/10.1107/97809553602060000727)
* [STAR file parser in Java](http://www.globalphasing.com/startools/)
* [C++ mmCIF reader/writer](http://sw-tools.pdb.org/apps/CIFPARSE-OBJ/index.html)
* [CIF syntax reference](http://www.iucr.org/resources/cif/spec/version1.1/cifsyntax)

**Comment by frodofine** [comment]
I've looked int the C++ mmCIF parser in the link above, but I feel the license is too restrictive and would conflict with Chemfile's current license.  Therefore, I've rolled my own version which is compatible with the files provided by RCSB, the COD, the CCSD, and PyMOL. I'll push this in a day or so.

**Comment by Luthaf** [comment]
Whow, that's great ! I started writting a CIF parser ([here](https://github.com/chemfiles/)), but never finished it ... 

Do you implement the whole CIF grammar ? I would like to use it for pure crystallographic CIF files too !

**Comment by frodofine** [comment]
Nope, I took the dirty PyMOL approach and have it recognize a very small set of the CIF grammar (just atom site at the moment). My application is the COD and CCSD, so I'm only testing against those databases (no idea if that's 'pure' or not!).

**Comment by fxcoudert** [comment]
I am interested in support for the CIF format as well.
Would a dependency on [GEMMI](https://gemmi.readthedocs.io/en/latest/) be acceptable here? 

- It boasts having the “ the fastest open-source CIF parser”
- Its CIF parser is [header-only](https://gemmi.readthedocs.io/en/latest/cif.html) so it's simple to integrate as a new dependency
- I have worked with it and can confirm it is good code, easy to use

**Comment by Luthaf** [comment]
I agree GEMMI looks nice and is pretty fast! I have two concerns with bringing it as a dependency:

- the license it uses (MPL) is not as permissive as the BSD we are using. It still is compatible
- using our own cif parser (https://github.com/chemfiles/cifxx) would allow us to use the same facilities for file reading as the rest of the code. My impression is that GEMMI takes either a `std::istream` or a `std::string`, meaning that for large files we will not have the full I/O niceties of our `TextFile` implementation.

---

At the same time, I have been meaning to work on this since 2015, and never actually got to it, in particular because I don't know how to deal properly with symmetry (which attribute from the CIF file should we use for space group detection? should we use `_symmetry_equiv_pos_as_xyz` instead? How should we deal with the same atoms having multiple images at the same  position? How should we deal with `_atom_site_occupancy` being different from 1.0?)

I guess a good way forward for now is to integrate GEMMI for now, possibly with a CMake option to opt-out to deal with license questions, and potentially replace it later with our own parser & symmetry code. Would you have time to do something like this?

**Comment by fxcoudert** [comment]
Let's try. For the record, I have already contributed a CIF file reader to VMD molfile (https://www.ks.uiuc.edu/Research/vmd/mailing_list/vmd-l/31552.html) but integration will be… slow, because VMD does not want C++11 dependencies. :sigh:

**Comment by Luthaf** [comment]
Then the quickest way to get this integrated in chemfiles would be to add the molfile plugin in https://github.com/chemfiles/molfiles/, and bind it in https://github.com/chemfiles/chemfiles/blob/7d744d67e8bc7245613534a2a303aac9d836eb5e/src/formats/Molfile.cpp. The resulting code would have all the limitations of a molfiles plugins (read only, constant number of atoms only in multi-frame file, high memory usage in multi-frame file), so you might still be interested in building a separate Format implementation. We have some documentation here http://chemfiles.org/chemfiles/latest/devdoc/format.html, and feel free to ask more questions!

**Comment by wojdyr** [comment]
> For the record, I have already contributed a CIF file reader to VMD molfile

wow, I didn't know about this, thanks!
If you'd need any help from the gemmi side feel free to contact me.

**Comment by fxcoudert** [comment]
@wojdyr hello! Gemmi is awesome, it's great to have such a small, clean implementation of CIF reading
